### ISSUE TITLE
`query_by_weights` doesn't seem to work as expected

### ISSUE THREAD

**Issue by vhpoet** [typesense/typesense#280]
## Description

I have a `persons` collection with `name`, `intro`, `weight` and some other (irrelevant) fields in the index. 

Now a few issues:

1. Somehow searching for "Bill" brings up Melinda Gates before Bill Gates even though the search should give a way higher weight to the `name` field.
2. Bill Gates has a way higher weight than Sonny Bill Williams, but Sonny is coming up first.

![Screen Shot 2021-05-31 at 3 53 33 PM](https://user-images.githubusercontent.com/686185/120247460-5c6c1680-c228-11eb-9c05-15be68c3bf51.png)

Search object:

```js
{
    collection: "persons",
    q: "Bill",
    query_by: "name, slug, twitter, intro",
    query_by_weights: "100, 2, 1, 1",
    per_page: 6,
    num_typos: 1,
    highlight_start_tag: "<em>",
    highlight_end_tag: "</em>",
    drop_tokens_threshold: 0,
}
```

Schema:
```js
{
  name: "persons",
  fields: [
    { name: "id", type: "string" },
    { name: "name", type: "string" },
    { name: "slug", type: "string" },
    { name: "weight", type: "int32" },
    { name: "imageSrc", type: "string", optional: true },
    { name: "twitter", type: "string", optional: true },
    { name: "intro", type: "string", optional: true },
  ],
  default_sorting_field: "weight",
};

```

Raw response from Typesense:

```js
"hits": [
    {
      "document": {
        "id": "246582a6-a716-4c26-9747-e7c827b73eb7",
        "name": "Sonny Bill Williams",
        "slug": "sonny-bill-williams",
        "twitter": "sonnybwilliams",
        "weight": 895223
      },
      "highlights": [
        {
          "field": "name",
          "matched_tokens": [
            "Bill"
          ],
          "snippet": "Sonny <em>Bill</em> Williams"
        }
      ],
      "text_match": 16907875
    },
    {
      "document": {
        "id": "b61db2b0-86a2-4e12-9022-a3eac26c3cc4",
        "intro": "Co-founder of Bill & Melinda Gates Foundation",
        "name": "Melinda Gates",
        "slug": "melinda-gates",
        "twitter": "melindagates",
        "weight": 2477539
      },
      "highlights": [
        {
          "field": "intro",
          "matched_tokens": [
            "Bill"
          ],
          "snippet": "Co-founder of <em>Bill</em> & Melinda Gates Foundation"
        }
      ],
      "text_match": 16882432
    },
    {
      "document": {
        "id": "8497255f-fbca-4736-84ca-8138a97a6fef",
        "intro": "Venture capitalist",
        "name": "Bill Gurley",
        "slug": "bill-gurley",
        "twitter": "bgurley",
        "weight": 432792
      },
      "highlights": [
        {
          "field": "name",
          "matched_tokens": [
            "Bill"
          ],
          "snippet": "<em>Bill</em> Gurley"
        }
      ],
      "text_match": 16857088
    },
    {
      "document": {
        "id": "b726967e-1bbe-4b67-999b-ba71fbee895b",
        "intro": "Co-founder and general partner at Craft",
        "name": "Bill Lee",
        "slug": "bill-lee",
        "twitter": "westcoastbill",
        "weight": 42910
      },
      "highlights": [
        {
          "field": "name",
          "matched_tokens": [
            "Bill"
          ],
          "snippet": "<em>Bill</em> Lee"
        }
      ],
      "text_match": 16857088
    },
    {
      "document": {
        "id": "0764f361-6523-4648-8703-2d524d602a42",
        "intro": "Co-founder of Microsoft",
        "name": "Bill Gates",
        "slug": "bill-gates",
        "twitter": "BillGates",
        "weight": 50063984
      },
      "highlights": [
        {
          "field": "name",
          "matched_tokens": [
            "Bill"
          ],
          "snippet": "<em>Bill</em> Gates"
        }
      ],
      "text_match": 16842752
    },
    {
      "document": {
        "id": "b8b09b1e-55fc-45a0-b4be-598bbb8c8e48",
        "intro": "42nd U.S. President",
        "name": "Bill Clinton",
        "slug": "bill-clinton",
        "twitter": "BillClinton",
        "weight": 11836504
      },
      "highlights": [
        {
          "field": "name",
          "matched_tokens": [
            "Bill"
          ],
          "snippet": "<em>Bill</em> Clinton"
        }
      ],
      "text_match": 16842752
    }
],
```

## Metadata

**Typsense Version**: 0.20.0
**OS**: CentOS

**Comment by kishorenc** [comment]
Thank you for reporting this issue. Do you still get the same order when you index just these 6 records and make the same query?

**Comment by vhpoet** [comment]
@kishorenc just created a new collection with just these 6 people and the result is the same.

Multisearch
```js
{
    "searches": [
    {
      "collection": "persons2",
      "q": "bill",
      "query_by": "name, slug, twitter, intro",
      "query_by_weights": "100, 2, 1, 1",
      "per_page": 6,
      "num_typos": 1,
      "highlight_start_tag": "<em>",
      "highlight_end_tag": "</em>",
      "drop_tokens_threshold": 0
    }
  ]
}
```

Response
```js
{
    "results": [
        {
            "facet_counts": [],
            "found": 6,
            "hits": [
                {
                    "document": {
                        "id": "246582a6-a716-4c26-9747-e7c827b73eb7",
                        "name": "Sonny Bill Williams",
                        "slug": "sonny-bill-williams",
                        "twitter": "sonnybwilliams",
                        "weight": 895223
                    },
                    "highlights": [
                        {
                            "field": "name",
                            "matched_tokens": [
                                "Bill"
                            ],
                            "snippet": "Sonny <em>Bill</em> Williams"
                        }
                    ],
                    "text_match": 16907875
                },
                {
                    "document": {
                        "id": "b61db2b0-86a2-4e12-9022-a3eac26c3cc4",
                        "intro": "Co-founder of Bill & Melinda Gates Foundation",
                        "name": "Melinda Gates",
                        "slug": "melinda-gates",
                        "twitter": "melindagates",
                        "weight": 2477539
                    },
                    "highlights": [
                        {
                            "field": "intro",
                            "matched_tokens": [
                                "Bill"
                            ],
                            "snippet": "Co-founder of <em>Bill</em> & Melinda Gates Foundation"
                        }
                    ],
                    "text_match": 16882432
                },
                {
                    "document": {
                        "id": "8497255f-fbca-4736-84ca-8138a97a6fef",
                        "intro": "Venture capitalist",
                        "name": "Bill Gurley",
                        "slug": "bill-gurley",
                        "twitter": "bgurley",
                        "weight": 432792
                    },
                    "highlights": [
                        {
                            "field": "name",
                            "matched_tokens": [
                                "Bill"
                            ],
                            "snippet": "<em>Bill</em> Gurley"
                        }
                    ],
                    "text_match": 16857088
                },
                {
                    "document": {
                        "id": "b726967e-1bbe-4b67-999b-ba71fbee895b",
                        "intro": "Co-founder and general partner at Craft",
                        "name": "Bill Lee",
                        "slug": "bill-lee",
                        "twitter": "westcoastbill",
                        "weight": 42910
                    },
                    "highlights": [
                        {
                            "field": "name",
                            "matched_tokens": [
                                "Bill"
                            ],
                            "snippet": "<em>Bill</em> Lee"
                        }
                    ],
                    "text_match": 16857088
                },
                {
                    "document": {
                        "id": "0764f361-6523-4648-8703-2d524d602a42",
                        "intro": "Co-founder of Microsoft",
                        "name": "Bill Gates",
                        "slug": "bill-gates",
                        "twitter": "BillGates",
                        "weight": 50063984
                    },
                    "highlights": [
                        {
                            "field": "name",
                            "matched_tokens": [
                                "Bill"
                            ],
                            "snippet": "<em>Bill</em> Gates"
                        }
                    ],
                    "text_match": 16842752
                },
                {
                    "document": {
                        "id": "b8b09b1e-55fc-45a0-b4be-598bbb8c8e48",
                        "intro": "42nd U.S. President",
                        "name": "Bill Clinton",
                        "slug": "bill-clinton",
                        "twitter": "BillClinton",
                        "weight": 11836504
                    },
                    "highlights": [
                        {
                            "field": "name",
                            "matched_tokens": [
                                "Bill"
                            ],
                            "snippet": "<em>Bill</em> Clinton"
                        }
                    ],
                    "text_match": 16842752
                }
            ],
            "out_of": 6,
            "page": 1,
            "request_params": {
                "collection_name": "persons2",
                "per_page": 6,
                "q": "bill"
            },
            "search_time_ms": 0
        }
    ]
}
```

**Comment by kishorenc** [comment]
Got it. I should then be able to zero in onto the issue easily with those records. I will get back to you once I've done that.

**Comment by kishorenc** [comment]
@vhpoet I'm able to reproduce this issue on 0.20.0 but it is fixed on the latest RC build. Can you please try against this Docker image: `typesense/typesense:0.21.0.rc10`.

If that looks fine, I can make a DEB/RPM build available for you.

**Comment by kishorenc** [comment]
@vhpoet Did you get a chance to try the RC build to see if it addresses the issue?

**Comment by vhpoet** [comment]
@kishorenc I can confirm that it works as expected now. An rpm package would be great, thank you!

**Comment by kishorenc** [comment]
Please use [this RPM](https://dl.typesense.org/releases/0.21.0.rc13/typesense-server-0.21.0.rc13-1.x86_64.rpm).

> Shouldn't Bill Gates be the first here too?

In this case, the term `gates` matches across two fields in the first result (both name and intro) but only against a single field in the second result (name). So this ordering is correct.

**Comment by vhpoet** [comment]
@kishorenc I understand that reasoning, but how can I give more weight to `weight`? Bill Gates has a way bigger weight and I'd like to see him come up first.

**Comment by kishorenc** [comment]
Can you try using the weight as the first sorting field like this: `sort_by=weight:desc,_text_match:desc`

**Comment by vhpoet** [comment]
@kishorenc that works.

**Comment by kishorenc** [comment]
I'm able to reproduce the issue and it is fixed in this Docker build: `typesense/typesense:0.21.0.rc15` -- if you can confirm, I can make a new RPM available.

**Comment by kishorenc** [comment]
Shipped in `0.21.0` release.
